### PR TITLE
Add an option to select individual lines for test generation

### LINKED ISSUE #12: Test generation for specific lines
# Description
 As a developer, I should be able to select specific lines in the IDE and tell the plugin to generate tests that cover those lines if they are valid (statements within a method).

Since EvoSuite only supports methods and classes for Units Under Test (UUT) but not specific lines, this feature would wrap around existing functionality and extend on it in the **plugin** layer.

# Definition of Done
* [x] The plugin is able to call EvoSuite with the relevant parameters so that the coverage goal is as likely as possible to be achieved.
* [x]  The user is not allowed to select invalid lines (like e.g declarations, empty lines)

### PR DESCRIPTION
**Pull request by jzelenjak**
# Description of changes made
Line selection for test generation is now supported. Invalid lines are not allowed to be selected.

# Why is merge request needed
This merge request is important since it adds the ability to select the lines for test generation which is part of an issue requested by the client.

# Other notes
Closes #12 

# What is missing?
- Dealing with the output tests generated by EvoSuite

- [x] I have checked that I am merging into correct branch

### CODE REVIEW

**Review by lyuben-todorov — DISMISSED** [review]

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        val projectClassPath = "$projectPath/target/classes/"
+
+        log.info("Generating tests for project $projectPath with classpath $projectClassPath")
+
+        log.info("Selected class is $classFQN")
+        log.info("Selected line is $selectedLine")
+
+        Runner(project, projectPath, projectClassPath, classFQN, fileName, modificationStamp).forLine(selectedLine)
+            .runEvoSuite()
+    }
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:51]: Since much of the code in this function is duplicated in all 3 test generation actions (class, method and line) is it possible to abstract it in some way, either by making a parent class or by moving it into a helper?

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt [hunk 2]
```diff
@@ -62,6 +67,28 @@ fun getSurroundingMethod(psiFile: PsiFile, caret: Caret): PsiMethod? {
...
+    val psiMethod: PsiMethod = getSurroundingMethod(psiFile, caret) ?: return null
+
+    val doc: Document = PsiDocumentManager.getInstance(psiFile.project).getDocument(psiFile) ?: return null
+
+    val selectedLine: Int = doc.getLineNumber(caret.offset)
+    val selectedLineText: String =
+        doc.getText(TextRange(doc.getLineStartOffset(selectedLine), doc.getLineEndOffset(selectedLine)))
+
+    if (selectedLineText.isBlank()) return null
+    return if (!validateLine(selectedLine, psiMethod, psiFile)) null else selectedLine
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt:89]: Can you maybe move this "if" outside of the return statement to a guard clause in order to improve readability?

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 1 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:51]: No

Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:51]: It would be useless since otherwise we will have to create methods with huge parameter lists or many return types. This will also not be good

Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:51]: At least i don't see a decent solution for it

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt (hunk 2 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt]: I thought it would be cool to use the fact that in Kotlin if-statement is an expression. But alright

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 1 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:51]: Because all the duplicated lines are just getting the stuff from the event

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/evosuite/SettingsArguments.kt [hunk 3]
```diff
@@ -33,12 +33,12 @@ class SettingsArguments(
...
-     * This makes EvoSuite create goals only for a certain method
-     * of the CUT.
+     * This makes EvoSuite create goals only for a certain method of the CUT.
      *
      * See https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.3
      *
      * @param method The descriptor of the method we're generating tests for
+     * @return the instance object
      */
     fun forMethod(method: String): SettingsArguments {
```
Comment (bolkhod) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/evosuite/SettingsArguments.kt:43]: What does this method do, I don't get it. It gets a string and returns `this`?

Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/evosuite/SettingsArguments.kt:43]: This is not my code, but i think it is a Builder pattern

**Review by kirilvasilev16 — APPROVED** [review]
Nice refactoring of the classes.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    override fun actionPerformed(e: AnActionEvent) {
+        val evoSuiteRunner: Runner = createEvoSuiteRunner(e) ?: return
+
+        val psiFile: PsiFile = e.dataContext.getData(CommonDataKeys.PSI_FILE) ?: return
+        val caret: Caret = e.dataContext.getData(CommonDataKeys.CARET)?.caretModel?.primaryCaret ?: return
+
+        val selectedLine: Int = getSurroundingLine(psiFile, caret)?.plus(1)
+            ?: return // lines in the editor and in EvoSuite are one-based
+
+        Logger.getInstance("GenerateTestsUtils").info("Selected line is $selectedLine")
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:31]: You should use a logger corresponding to this class (e.g. `Logger.getInstance(this.javaClass)`)

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt [hunk 5]
```diff
@@ -19,50 +16,30 @@ import nl.tudelft.ewi.se.ciselab.testgenie.helpers.generateMethodDescriptor
  *   getting the information about the selected method and passing it to (EvoSuite) Runner.
  */
 class GenerateTestsActionMethod : AnAction() {
-    private val log = Logger.getInstance(this.javaClass)
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt:22]: Why did you remove this logger?

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt [hunk 6]
```diff
@@ -19,50 +16,30 @@ import nl.tudelft.ewi.se.ciselab.testgenie.helpers.generateMethodDescriptor
...
-        val classFQN = containingClass.qualifiedName ?: return
         val methodDescriptor = generateMethodDescriptor(psiMethod)
 
-        val projectPath: String = ProjectRootManager.getInstance(project).contentRoots.first().path
-        val projectClassPath = "$projectPath/target/classes/"
-
-        log.info("Generating tests for project $projectPath with classpath $projectClassPath")
-
-        log.info("Selected class is $classFQN, method is $methodDescriptor")
+        Logger.getInstance("GenerateTestsUtils").info("Selected method is $methodDescriptor")
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt:33]: You should use the logger you removed above.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt [hunk 7]
```diff
@@ -62,6 +103,30 @@ fun getSurroundingMethod(psiFile: PsiFile, caret: Caret): PsiMethod? {
...
+/**
+ * Gets the selected line if the constraints on the selected line are satisfied,
+ *   so that EvoSuite can generate tests for it.
+ * Namely, the line is within a method, it is not blank and contains (part of) a statement.
+ *
+ * @param psiFile the current PSI file (where the user makes the click)
+ * @param caret the current (primary) caret that did the click
+ * @return line number if the constraints are satisfied else null
+ */
+fun getSurroundingLine(psiFile: PsiFile, caret: Caret): Int? {
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt:115]: Maybe this method should be part of the GenerateTestsActionLine instead of here since it's only used there.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt [hunk 8]
```diff
@@ -122,6 +187,30 @@ private fun validateClass(psiClass: PsiClass): Boolean {
...
+/**
+ * Checks if the selected line contains (part of) a statement and is a valid line to be selected for EvoSuite.
+ * Namely, the line is within a method, it is not blank and contains (part of) a statement.
+ *
+ * @param selectedLine selected line number
+ * @param psiMethod surrounding PSI method
+ * @param psiFile containing PSI file
+ * @return true if the line is valid, false otherwise
+ */
+private fun validateLine(selectedLine: Int, psiMethod: PsiMethod, psiFile: PsiFile): Boolean {
```
Comment (martinmladenov) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt:199]: Same for this one

File: src/test/resources/pizzeria/src/PizzaClasse.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+interface PizzaServizio {
+    default void saluto() {
+        System.out.println("Viva la pizza Italiana");
+    }
+
+    PizzaAstratta fareLaPizza();
+}
+
+abstract class PizzaAstratta {
```
Comment (martinmladenov) [inline src/test/resources/pizzeria/src/PizzaClasse.java:40]: Questa è una pizza molto buona!

**Review by martinmladenov — COMMENTED** [review]
This is a very nice refactoring! It definitely improves code quality a lot.

I just have a couple of small comments and it's perfect!

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 1 above)
Comment (martinmladenov) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt]: Great, thanks!

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 4 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:31]: The problem is that GenerateTestsUtils is not a class, hence i cannot do this trick with it. So, for consistency, i named all of them as "GenerateTestsUtils"

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt (hunk 6 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt:33]: Why? I don't think it is necessary to store it as a separate private variable now that it prints only one line. At least to me it seems like an overhead. If i understand correctly what you mean

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt (hunk 7 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt:115]: Similar to `getSurroundingClass` and `getSurroudingMethod`, i have decided to put them here to keep *GenerateTestsAction* classes only for `actionPerformed` and `update` methods (which is why these classes exist). I think it would be more clutter if we add all those methods to the aciton class. Also, this method might be potentially used in other cases (maybe by future developers/contributors? ) and it is not something that is inherent to action class.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 4 above)
Comment (martinmladenov) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:31]: But this is in the GenerateTestsActionLine class, which is a class. You can do this here.

You shouldn't instantiate loggers named after a different class since that creates confusion.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt (hunk 8 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt:199]: Same comment as above. Also, it would defeat the purpose of GenerateTestsUtils class. I think that such business logic has to be separated into a separate class, and action classes have to be responsible for giving (showing) the option to generate tests to the user (when appropriate), and react to the action

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt (hunk 6 above)
Comment (martinmladenov) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt:33]: I meant the name of the logger - this logger is named after GenerateTestsUtils but you use it in the GenerateTestsActionMethod.

(Also, creating a private variable won't create any overhead, don't worry about that)

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt (hunk 7 above)
Comment (martinmladenov) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsUtils.kt:115]: Sure, fair point

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 4 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt:31]: How do i name the one for GenerateTestsUtils?

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt (hunk 6 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt:33]: Is it better to create one or not? I would not, personally

Comment (martinmladenov) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt:33]: I think it would be better to create a private variable since it allows for easier extensibility when you want to modify this class and add another log message.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt (hunk 4 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionLine.kt]: Done

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt (hunk 6 above)
Comment (jzelenjak) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/actions/GenerateTestsActionMethod.kt]: Done

**Review by martinmladenov — APPROVED** [review]
Thanks for addressing the comments, LGTM!

**Review by kirilvasilev16 — APPROVED** [review]
